app/src/main/java/com/rkm/tasky/network/authentication/implementation/LoginManagerImpl.kt

```diff
@@ -0,0 +1,69 @@
+package com.rkm.tasky.network.authentication.implementation
+
+import com.rkm.tasky.di.IoDispatcher
+import com.rkm.tasky.network.authentication.abstraction.LoginManager
+import com.rkm.tasky.network.model.request.AccessTokenRequest
+import com.rkm.tasky.network.model.request.LoginUserRequest
+import com.rkm.tasky.network.model.response.asAuthInfo
+import com.rkm.tasky.network.repository.abstraction.TaskyRemoteRepository
+import com.rkm.tasky.util.storage.implementation.SessionStorageImpl
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class LoginManagerImpl @Inject constructor(
+    private val repository: TaskyRemoteRepository,
+    private val sessionStorage: SessionStorageImpl,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+) : LoginManager {
+    override suspend fun logIn(email: String, password: String): Result<Boolean> =
+        withContext(dispatcher) {
+            val result = repository.loginUser(LoginUserRequest(email, password))
+            return@withContext if (result.isSuccess) {
+                sessionStorage.setSession(result.getOrNull()!!.asAuthInfo())
+                Result.success(true)
+            } else {
+                Result.failure(result.exceptionOrNull()!!)
```

Comment:
With the above generic exception handler, you can avoid these double-bang (`!!`) operators

```
val result = api.login(…) // <— Returns a `Result<AuthInfo, DataError>`
when(result) {
  is Result.Success -> // safely access the `AuthInfo` object 
  is Result.Error -> // handle failure case with access to the `DataError` as the cause
}
```